### ISSUE TITLE
no-undef doesn't fire on self reference naming

### ISSUE THREAD

**Issue by benjick** [eslint/eslint#7333]
**Tell us about your environment**

* **ESLint Version:** 3.7.1
* **Node Version:** v6.3.0
* **npm Version:** 3.10.3

**What parser (default, Babel-ESLint, etc.) are you using?**
Default

**Please show your full configuration:**

```
{
  "rules": {
    "no-undef": 1
  }
}

```

**What did you do? Please include the actual source code causing the issue.**

```js
var bar = quz;
var foo = foo;
```

**What did you expect to happen?**
```
eslint-no-undef-self-ref/index.js
  1:11  warning  'quz' is not defined  no-undef
  2:11  warning  'foo' is not defined  no-undef

✖ 2 problems (0 errors, 2 warnings)
```

**What actually happened? Please include the actual, raw output from ESLint.**
```
eslint-no-undef-self-ref/index.js
  1:11  warning  'quz' is not defined  no-undef

✖ 1 problem (0 errors, 1 warning)
```

Repo with reproduction: https://github.com/benjick/eslint-no-undef-self-ref

**Comment by not-an-aardvark** [comment]
Hello, thanks for the report.

I think this is intended behavior; to prevent this, you should use [`no-use-before-define`](http://eslint.org/docs/rules/no-use-before-define) instead.

See also: https://github.com/eslint/eslint/issues/5437, https://github.com/eslint/eslint/issues/7111

**Comment by benjick** [comment]
This also happens with const, which aren't hoisted, right? 

Thanks for the reply

**Comment by not-an-aardvark** [comment]
Yes, it does. I think the consensus on that issue was to still delegate that case to `no-use-before-define` rather than `no-undef`.

**Comment by vitorbal** [comment]
Looks like there's nothing left to do here, as it seems the questions have been answered. Closing this, but feel free to continue the discussion if needed.
